src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/FunctionDescriptor.java

```diff
@@ -123,10 +123,27 @@ public int firstVariadicArgumentIndex() {
      * @param addedLayouts the argument layouts to append.
      * @return the new function descriptor.
      */
-    public FunctionDescriptor withAppendedArgumentLayouts(MemoryLayout... addedLayouts) {
+    public FunctionDescriptor appendArgumentLayouts(MemoryLayout... addedLayouts) {
+        return insertArgumentLayouts(argLayouts.size(), addedLayouts);
+    }
+
+    /**
+     * Create a new function descriptor with the given argument layouts inserted at the given index, into the argument
+     * layout array of this function descriptor.
+     * @param index the index at which to insert the arguments
+     * @param addedLayouts the argument layouts to insert at given index.
+     * @return the new function descriptor.
+     * @throws IllegalArgumentException if {@code index < 0 || index > argumentLayouts().size()}.
+     */
+    public FunctionDescriptor insertArgumentLayouts(int index, MemoryLayout... addedLayouts) {
         Objects.requireNonNull(addedLayouts);
         Arrays.stream(addedLayouts).forEach(Objects::requireNonNull);
```

Comment:
FWIW you can remove this line if you wish, since the `List.of()` will check for null elements. You could replace lines 139 to 143 with:
```java
if (index < 0 || index > argLayouts.size())
    throw new IllegalArgumentException("Index out of bounds: " + index);
List<MemoryLayout> added = List.of(addedLayouts); // null check on array and its elements
List<MemoryLayout> newLayouts = new ArrayList<>(argLayouts.size() + added.size());
```